### PR TITLE
apiserver: move to dependency engine

### PR DESCRIPTION
**Pull request by axw**
## Description of change

Move the API server and mongo upgrader workers to the
dependency engine. The API server now uses the State
provided by the state-tracker manifold. Later we will
extend the state-tracker to provide a StatePool, at
which point we can stop constructing one in the API
server manifold.

The API server no longer runs a worker to ping the
State's underlying MongoDB session, since that is
done by the state-tracker worker.

Also, introduce another manifold for watching agent
config changes, extracting and providing the TLS
certificate for the API server. This replaces the
"CertChanged" channel that was previously provided
to the API server.

## QA steps

This is just a lot of shuffling around, so just smoke test by bootstrapping and deploying a couple of charms, enable-ha, and run "juju upgrade-juju -m controller --build-agent".

## Documentation changes

None.

## Bug reference

None.

### CODE REVIEW

File: worker/apiservercertwatcher/manifold_test.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+	cert_ := f()
+	c.Assert(cert, gc.Equals, cert_)
+}
+
+func (s *ManifoldSuite) TestCertUpdated(c *gc.C) {
+	w := s.startWorkerClean(c)
+	defer workertest.CleanKill(c, w)
+
+	var f func() *tls.Certificate
```
Comment (wallyworld) [inline worker/apiservercertwatcher/manifold_test.go]: maybe s/f/getCert for clarity? YMMV.

File: worker/mongoupgrader/manifold.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package mongoupgrader
```
Comment (wallyworld) [inline worker/mongoupgrader/manifold.go]: All of this mongo upgrade stuff
- the worker
- the facade apis
- the machine doc attribute

... none of it is used. There's an upgrade-mongo plugin which is the only thing which calls the facade api, and we are dropping the plugin as it is not used. It was only for back in the 1.26 days when we were going to upgrade mongo. 

As a first step to getting of this stuff. let's drop the worker from this PR. A followup can remove the facades and other bits.

File: worker/apiservercertwatcher/manifold_test.go (hunk 1 above)
Comment (axw) [reply worker/apiservercertwatcher/manifold_test.go]: ok

File: worker/mongoupgrader/manifold.go (hunk 2 above)
Comment (axw) [reply worker/mongoupgrader/manifold.go]: ok, added a commit to drop the worker and manifold

**Comment by axw** [comment]
[Move API server to dependency engine](https://trello.com/c/jyaTb8de/15-move-api-server-to-dependency-engine)

**Comment by axw** [comment]
> Can we also test that Juju will correctly responds to a controller being taken down in HA mode, and that a subsequent enable-ha causes things to readjust.

Done.

**Comment by axw** [comment]
$$merge$$

**Pull request merged**